### PR TITLE
Feature/resolve page path

### PR DESCRIPTION
**Pull request by willcAND**
### What

Implemented page path logic in conjunction with diagram 3

### How to review

Please review the diagram referenced in DIS-413 to verify that this Pull Request (PR) follows its specifications.

### Who can review

A member of the ONS

### CODE REVIEW

File: response/page_path.go [hunk 1]
```diff
@@ -2,13 +2,127 @@ package response
...
+var methodologiesOrQMIsOrAdhocsPathRegexp = regexp.MustCompile(`.+/(methodologies|qmis|adhocs)/([^/]+)`)
+var endsWithFileExtRegexp = regexp.MustCompile(`/[^/]+\.\w+$`)
+var timeSeriesPathRegexp = regexp.MustCompile(`^(.*?/timeseries(?:/[^/]+){0,2})`)
+var datasetsPathRegexp = regexp.MustCompile(`^(.*?/datasets(?:/[^/]+){0,2})`)
+
 func getPagePath(ctx context.Context, uri string) string {
 	log.Info(ctx, "Calculating page path for "+uri)
+	uri = strings.TrimSuffix(uri, "/")
+
+	if pagePath, isVisualisationsEndpoint := resolveVisualisationsEndpoint(uri); isVisualisationsEndpoint {
```
Comment (melsner11Ons) [inline response/page_path.go:26]: check for "ons" is missing before visualisation?

File: response/page_path.go [hunk 2]
```diff
@@ -2,13 +2,127 @@ package response
...
 
+var visualisationsEndpointRegexp = regexp.MustCompile(`(^/visualisations/.+?)/`)
+var bulletinsOrArticlesPathRegexp = regexp.MustCompile(`(.+/(bulletin|article)s/.+?/[^/]+)`)
+var methodologiesOrQMIsOrAdhocsPathRegexp = regexp.MustCompile(`.+/(methodologies|qmis|adhocs)/([^/]+)`)
+var endsWithFileExtRegexp = regexp.MustCompile(`/[^/]+\.\w+$`)
+var timeSeriesPathRegexp = regexp.MustCompile(`^(.*?/timeseries(?:/[^/]+){0,2})`)
+var datasetsPathRegexp = regexp.MustCompile(`^(.*?/datasets(?:/[^/]+){0,2})`)
+
 func getPagePath(ctx context.Context, uri string) string {
 	log.Info(ctx, "Calculating page path for "+uri)
```
Comment (melsner11Ons) [inline response/page_path.go:22]: maybe add a check for null in uri to be safe? Also maybe trim leading and trailing whitespace?

File: response/page_path_test.go [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+type TestCase struct {
+	uri      string
+	expected string
+}
+
+func TestGetPagePath(t *testing.T) {
+	Convey("Given a list of URIs", t, func() {
+		ctx := context.Background()
+
+		visualisationsEndpoints := []TestCase{
```
Comment (melsner11Ons) [inline response/page_path_test.go:19]: maybe add 2 test for very unlikely paths : empty string -> "", slash only "/", two slashes "//", and null

File: response/page_path_test.go [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+			{
```
Comment (melsner11Ons) [inline response/page_path_test.go:20]: its very unlikely but i could imagine you get a path like "http://www.ons.gov.uk/eonomy/..." in that case it should remove the protocol and host and port

File: response/page_path.go (hunk 2 above)
Comment (melsner11Ons) [inline response/page_path.go:22]: does it make sense to convert the incoming string to lowercase? Not sure the matchers are case insensitive?

File: response/page_path_test.go [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+				expected: "/aboutus/whatwedo/programmesandprojects/sustainabledevelopmentgoals",
+			},
+			{
+				uri:      "/generator?format=csv&uri=/economy/grossvalueaddedgva/timeseries/abml/pn2/previous/v29",
+				expected: "/economy/grossvalueaddedgva/timeseries/abml/pn2",
+			},
+		}
+		resourceEndPointPathsWithEncodedURI := []TestCase{
+			{
+				uri:      "/generator?uri=%2Feconomy%2Feconomicoutputandproductivity%2Foutput%2Fbulletins%2Feconomicactivityandsocialchangeintheukrealtimeindicators%2F15february2024%2F426e63a0%26format%3Dcsv",
```
Comment (jmgq) [inline response/page_path_test.go]: This one does not look right, as it seems like the & from the `format` query param has also been encoded. It's only the value for the `uri` path param that should be encoded. This is not a real use case. It's still valuable, so if you remove the encoding from anything that's not the value of the `uri` param, then that should make it look like a real use case.

File: response/page_path_test.go [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+			{
+				uri:      "/visualisations/dvc1945/seasonalflu/index.html",
+				expected: "/visualisations/dvc1945",
+			},
+			{
+				uri:      "/visualisations/dvc1945/duetoinvolving/fallback.png",
+				expected: "/visualisations/dvc1945",
+			},
+		}
+		resourceEndPointPaths := []TestCase{
```
Comment (jmgq) [inline response/page_path_test.go]: We're missing some real examples of other endpoints, like `/embed` or `/chart` (and others). We should ask someone at ONS, so that they can tell us how to find the missing ones, and add them here.

File: response/page_path.go (hunk 1 above)
Comment (jmgq) [reply response/page_path.go:26]: No, that's not needed. If we make it to here, then `/ons` has already been checked in a previous step.

File: response/page_path.go (hunk 2 above)
Comment (jmgq) [reply response/page_path.go:22]: It's good that you're thinking about edge cases :-) But `uri` cannot be `nil`, it's always a string. Trimming is not necessary either, because this comes directly from the request, so it will be trimmed already.

File: response/page_path_test.go (hunk 3 above)
Comment (jmgq) [reply response/page_path_test.go:19]: `nil` is impossible, but it would be a good idea to have another list of test cases for the behaviour when none of the previous rules apply: to return the URI unmodified. It would also be good to add some real URLs there.

File: response/page_path_test.go (hunk 4 above)
Comment (jmgq) [reply response/page_path_test.go:20]: This cannot happen, we get the path from the request struct, so the host and protocol are never part of it.

File: response/page_path.go (hunk 2 above)
Comment (jmgq) [reply response/page_path.go:22]: No, that would be wrong. It's only the protocol and domain that are case-insensitive, but everything after that is case-sensitive. `/test` and `/TEST` are two different paths.

File: response/page_path.go [hunk 7]
```diff
@@ -2,13 +2,161 @@ package response
...
 
 	"github.com/ONSdigital/log.go/v2/log"
 )
 
-func getPagePath(ctx context.Context, uri string) string {
+var resourceEndpoints = []string{"/chartconfig", "/chartimage", "/embed", "/chart", "/resource", "/generator", "/file", "/export"}
+
+var visualisationsEndpointRegexp = regexp.MustCompile(`(^/visualisations/.+?)/`)
+var bulletinsOrArticlesPathRegexp = regexp.MustCompile(`(.+/(bulletin|article)s(?:/[^/]+){2})`)
+var methodologiesOrQMIsOrAdHocsPathRegexp = regexp.MustCompile(`.+/(methodologie|qmi|adhoc)s/([^/]+)`)
```
Comment (gedge) [inline response/page_path.go]:
```suggestion
var bulletinsOrArticlesPathRegexp = regexp.MustCompile(`(.+/(bulletins|articles)(?:/[^/]+){2})`)
var methodologiesOrQMIsOrAdHocsPathRegexp = regexp.MustCompile(`.+/(methodologies|qmis|adhocs)/([^/]+)`)
```

make these more clear/searchable (i.e. if we want to search our code for `methodologies`, this makes it easier)

File: response/page_path.go [hunk 8]
```diff
@@ -2,13 +2,161 @@ package response
...
+	"regexp"
+	"strings"
 
 	"github.com/ONSdigital/log.go/v2/log"
 )
 
-func getPagePath(ctx context.Context, uri string) string {
+var resourceEndpoints = []string{"/chartconfig", "/chartimage", "/embed", "/chart", "/resource", "/generator", "/file", "/export"}
+
+var visualisationsEndpointRegexp = regexp.MustCompile(`(^/visualisations/.+?)/`)
```
Comment (gedge) [inline response/page_path.go]:
```suggestion
var visualisationsEndpointRegexp = regexp.MustCompile(`(^/visualisations/[^/]+)/`)
```
why not use this pattern (in common with other regexp's here)?

File: response/page_path_test.go [hunk 9]
```diff
@@ -0,0 +1,356 @@
...
+				uri:      "/economy/inflationandpriceindices/datasets/consumerpriceindices/current",
+				expected: "/economy/inflationandpriceindices/datasets/consumerpriceindices/current",
+			},
+			{
+				uri:      "/employmentandlabourmarket/peopleinwork/earningsandworkinghours/datasets/analysisofwageandpriceincreases",
+				expected: "/employmentandlabourmarket/peopleinwork/earningsandworkinghours/datasets/analysisofwageandpriceincreases",
+			},
+			{
+				uri:      "/businessindustryandtrade/changestobusiness/mergersandacquisitions/datasets/timeseries/15march2024",
+				expected: "/businessindustryandtrade/changestobusiness/mergersandacquisitions/datasets/timeseries/15march2024",
```
Comment (gedge) [inline response/page_path_test.go:308]:
```suggestion
				uri:      "/businessindustryandtrade/changestobusiness/mergersandacquisitions/datasets/timeseries/15march2024/extra-third-segment",
				expected: "/businessindustryandtrade/changestobusiness/mergersandacquisitions/datasets/timeseries/15march2024",
```
should there be a test for the removal of `extra-third-segment`?

**Review by gedge — COMMENTED** [review]
hopefully small comments

File: response/page_path_test.go (hunk 9 above)
Comment (jmgq) [reply response/page_path_test.go:308]: The URIs you see in this group are real examples. We couldn't find a real example with 3 segments, so I've added a new test case with fake segment names, so that it's clear that even though it's not a real URI, we'd like to test that scenario 👍

**Review by gedge — APPROVED** [review]
LGTM
